### ISSUE TITLE
Using type alias in spec leads to failure of type checking

### ISSUE THREAD

**Issue by skogsbaer** [etylizer/etylizer#155]
This example works:

```erlang
-module(test).

-compile(export_all).
-compile(nowarn_export_all).

-spec maps_to_list(#{Key => Value}) -> [{Key, Value}].
maps_to_list(M) -> maps_to_list(M).

-spec range(#{ any() => V }) -> [V].
range(Map) ->
    lists:map(fun({_, V}) -> V end, maps_to_list(Map)).
```

But when making the spec of `maps_to_list` more similar to `maps:to_list` (because of #154, we cannot use the real spec), type checking fails:

```erlang
-module(test).

-compile(export_all).
-compile(nowarn_export_all).

-spec maps_to_list(Map) -> [{Key, Value}] when Map :: #{Key => Value}.
maps_to_list(M) -> maps_to_list(M).

-spec range(#{ any() => V }) -> [V].
range(Map) ->
    lists:map(fun({_, V}) -> V end, maps_to_list(Map)).
```

The error is:

```
[NOTE 2024-10-25 06:44:32.053 typing_check.erl:29] Checking failed: test.erl:11:18: Type error: this branch never matches
%   11|     lists:map(fun({_, V}) -> V end, maps_to_list(Map)).
%     |                  ^
test.erl:11:18: Type error: this branch never matches
%   11|     lists:map(fun({_, V}) -> V end, maps_to_list(Map)).
%     |
```

**Comment by albsch** [comment]
When I define the spec like this it type checks (i.e. expanded `Map :: #{Key => Value}`):

```
-spec maps_to_list(#{Key => Value}) -> [{Key, Value}].
```

**Comment by skogsbaer** [comment]
Just verified: the bug is still in main

**Comment by skogsbaer** [comment]
@albsch I'm pretty sure that this is again some kind of caching bug.

For the two examples, we have the constraint set

```
[{scsubty, 11:18, {$20}, not({{any(), any()}} /\ {any()})},{scsubty, 11:18, $7, $20}]
```

is not satisfiable, so the redundancy check in the first example does not fail.

But the constraint set

```
[{scsubty, 11:18, $8, $21}, {scsubty, 11:18, {$21}, not({{any(), any()}} /\ {any()})}]
```

is satisfiable, so the redundancy check in the second example fails.


But the two constraint sets are equal modulo ordering and variable renaming!

**Comment by albsch** [comment]
The constraint sets are different though:

Set 1:

```erlang
[
 {scsubty, 11:37, [{$5, $6}], $3 /\ [A]},
 {scsubty, 11:50, #{any() => V}, #{$5 => $6}},
...
 ]
```

Set 2:

```erlang
[
 {scsubty, 11:37, [{$6, $7}], $3 /\ [A]},
 {scsubty, 11:50, #{any() => V}, $5 /\ #{Key => Value}},
 ...
 ]
```

In set 2, the variable `$5` is not connected with any other constraint, neither is `Key` nor `Value`. In set 1, variables `$5` and `$6` are connected in two constraint sets.


Here are the full constraint sets (sorted):

```erlang
[
 {scsubty, 11:5, $4 /\ [B], [V]},
 {scsubty, 11:37, [{$5, $6}], $3 /\ [A]},
 {scsubty, 11:50, #{any() => V}, #{$5 => $6}},
 {scsubty, 11:30, $19, $8},
 {scsubty, 11:30, $14, $19},
 {scsubty, 11:18, $7, $10},
 {scsubty, 11:18, {$10}, $9},
 {scsubty, 11:18, $12, {$13, $14}},
 {scsubty, 11:18, $16, {$17, $18}},
 {scsubty, 11:18, $9, {{any(), any()}} /\ {any()}},
 {scsubty, 11:18, $9 /\ {{any(), any()}} /\ {any()}, {$16}},
 {scsubty, 11:18, $9 /\ {{any(), any()}} /\ {any()}, {$15}},
 {scsubty, 11:18, $9 /\ {{any(), any()}} /\ {any()}, {$12}},
 {scsubty, 11:18, $9 /\ {{any(), any()}} /\ {any()}, {$11}}
 {scsubty, 11:15, fun(($7) -> $8), $0 /\ fun((A) -> B)},
 ]

[
 {scsubty, 11:5, $4 /\ [B], [V]},
 {scsubty, 11:37, [{$6, $7}], $3 /\ [A]},
 {scsubty, 11:50, #{any() => V}, $5 /\ #{Key => Value}},
 {scsubty, 11:30, $20, $9},
 {scsubty, 11:18, $8, $11},
 {scsubty, 11:30, $15, $20},
 {scsubty, 11:18, {$11}, $10},
 {scsubty, 11:18, $13, {$14, $15}},
 {scsubty, 11:18, $17, {$18, $19}}
 {scsubty, 11:18, $10, {{any(), any()}} /\ {any()}},
 {scsubty, 11:18, $10 /\ {{any(), any()}} /\ {any()}, {$17}},
 {scsubty, 11:18, $10 /\ {{any(), any()}} /\ {any()}, {$16}},
 {scsubty, 11:18, $10 /\ {{any(), any()}} /\ {any()}, {$13}},
 {scsubty, 11:18, $10 /\ {{any(), any()}} /\ {any()}, {$12}},
 {scsubty, 11:15, fun(($8) -> $9), $0 /\ fun((A) -> B)},
 ]
```

**Comment by albsch** [comment]
Check against cduce, same result:

```
debug tallying (['a 'b] ['a 'b] [( 'a1, 'b)( 'a, 'a0)( 'a -> 'b, 'a0 -> 'a1 )]);;
Result:[{'a1:='b; 'a0:='a}]
debug tallying (['xb 'a 'b] ['xb 'a 'b] [( 'a1, 'b & 'xb)( 'a, 'a0)( 'a -> 'b, 'a0 -> 'a1 )]);;
Result:[]
```

**Comment by albsch** [comment]
Constraints that pass:

```erlang
[{scsubty, 8:16, $1, B},
 {scsubty, 8:16, fun((A) -> B), fun(($0) -> $1)},
 {scsubty, 8:20, A, $0}]
```

Constraints that fail:

```erlang
[{scsubty, 12:17, $1, Xb /\ B},
 {scsubty, 12:17, fun((A) -> B), fun(($0) -> $1)},
 {scsubty, 12:21, A, $0}]
```

Any ideas what the issue could be @skogsbaer ?

**Comment by skogsbaer** [comment]
@albsch Good analysis! I think we should expand such aliases before type checking, as suggested in EEP-71 (https://www.erlang.org/eeps/eep-0071). The EEP suggest to perform the expansion during parsing. This has the advantage that the type checker only sees types where such aliases have already been expanded. The downside is that error messages no longer contain the original but the expanded type spec. I think we can accept this behavior and take the simple route.

**Comment by skogsbaer** [comment]
I looked more carefully at this example.

@albsch gave the following "minimal example":

```erlang
-module(alias).

-spec id(fun((A) -> B), A) -> B.
id(Fun, Xa) -> Fun(Xa).

-spec id2(fun((A) -> B), A) -> Xb when Xb :: B.
id2(Fun, Xa) -> Fun(Xa).
```

Currently, `id` typechecks but `id2` does not. This is expected because we are (still) interpreting `::` as a subtype constraint. Then, `id2` should not typecheck, because the result of `id2` cannot be `Xb`.

**Comment by skogsbaer** [comment]
Here is again the problematic example:

```
-spec to_list(MapOrIterator) -> [{Key, Value}] when MapOrIterator :: #{Key => Value}.
to_list(X) -> to_list(X).

-spec range(#{ any() => V }) -> [V].
range(Map) ->
    lists:map(fun({_, V}) -> V end, to_list(Map)).
```

This example fails to type check. The error is:

```
./test_files/tycheck/range.erl:12:18: Type error: this branch never matches
%   12|     lists:map(fun({_, V}) -> V end, to_list(Map)).
%     |
```

To check for redundant branches, we first type check the function (`range` in this example). Then we check satisfiability of a specific constraint set. If the constraints are satisifiable, then the branch is redundant, otherwise it is not. For the example shown above, the constraints are:

```erlang
[{scsubty, 12:18, $8, $21}, {scsubty, 12:18, {$21}, not({{any(), any()}} /\ {any()})}]
```

with `fixed = [V]`

This constraint set is satisfiable, as shown cduce.

We now modify the example, arriving at:

```erlang
-spec to_list(#{Key => Value}) -> [{Key, Value}].
to_list(X) -> to_list(X).

-spec range(#{ any() => V }) -> [V].
range(Map) ->
    lists:map(fun({_, V}) -> V end, to_list(Map)).
```

With the modified example, the constraint set for checking satisifiability is the same, except for reordering and renaming of variables:

```
[{scsubty, 12:18, $7, $20}, {scsubty, 12:18, {$20}, not({{any(), any()}} /\ {any()})}]
```

with `fixed = [V]`

But now tally says that the constraint set is not satisfiable.

I think there are two bugs:

1. The constraint set for checking redundant branches seems to be wrong. I'm investigating.
2. tally returns different results for the same constraint set (module variable renaming and reordering of constraints). @albsch can you check?

**Comment by skogsbaer** [comment]
I delete a previous comment because it was not correct.

**Comment by skogsbaer** [comment]
Fixed with #176
